docs/templates/dependencies.md

```diff
@@ -0,0 +1,113 @@
+# Template Dependencies
+
+When creating Coder templates, it is unlikely that you will just be using
+built-in providers. Part of Terraform's flexibility stems from its rich plugin
+ecosystem, and it makes sense to take advantage of this.
+
+That having been said, here are some recommendations to follow, based on the
+[Terraform documentation](https://developer.hashicorp.com/terraform/tutorials/configuration-language/provider-versioning).
+
+Following these recommendations will:
+
+- **Prevent unexpected changes:** Your templates will use the same versions of
+  Terraform providers each build. This will prevent issues related to changes in
+  providers.
+- **Improve build performance:** Coder caches provider versions on each build.
+  If the same provider version can be re-used on subsequent builds, Coder will
+  simply re-use the cached version if it is available.
```

Comment:
We could explicitly state that cache utilization is only possible when a lock-file is present. I could be misremembering, but providers will always be re-downloaded otherwise, even if cached.